device/src/region/fixed_channel_plans/mod.rs

```diff
@@ -2,21 +2,170 @@ use super::*;
 mod us915;
 
 pub(crate) use au915::AU915;
 pub(crate) use us915::US915;
 
+seq_macro::seq!(
+    N in 0..=71 {
+        #[derive(Debug, Clone, Copy, PartialEq, Eq)]
+        #[repr(u8)]
+        pub enum Channel {
+            #(
+                _~N,
+            )*
+        }
+    }
+);
+
+impl From<Channel> for u8 {
+    fn from(value: Channel) -> Self {
+        value as u8
+    }
+}
+
+#[derive(Clone)]
+struct PreferredJoinChannels {
+    channel_list: heapless::Vec<u8, 16>,
+    // Number representing the maximum number of retries allowed using the preferred channels list,
+    max_retries: usize,
+    // Decrementing number representing how many tries we have left with the specified list before
+    // reverting to the default channel selection behavior.
+    num_retries: usize,
+}
+
+impl PreferredJoinChannels {
+    fn try_get_channel(&mut self, rng: &mut impl RngCore) -> Option<u8> {
+        if self.num_retries > 0 {
+            let random = rng.next_u32();
+            self.num_retries -= 1;
+            let len = self.channel_list.len();
+            // TODO non-compliant because the channel might be the same as the previously
+            // used channel?
+            Some(self.channel_list[random as usize % len])
+        } else {
+            None
+        }
+    }
+}
+
+/// Bitflags containing subbands that haven't yet been tried for a join attempt
+/// this round.
+#[derive(Clone)]
+struct AvailableSubbands(u16);
+
+impl AvailableSubbands {
+    const ALL_ENABLED: u16 = 0b111111111;
+
+    fn new() -> Self {
+        Self(Self::ALL_ENABLED)
+    }
+
+    fn is_empty(&self) -> bool {
+        self.0 == 0
+    }
+
+    fn pop_next(&mut self) -> Option<u8> {
+        for bit in 0..=8 {
+            if (self.0 >> bit) & 1 == 1 {
+                self.0 &= !(1 << bit);
+                return Some(bit);
+            }
+        }
+        None
+    }
+
+    fn reset(&mut self) {
+        self.0 = Self::ALL_ENABLED;
+    }
+}
+
+impl Default for AvailableSubbands {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+#[derive(Default, Clone)]
+struct JoinChannels {
+    preferred_channels: Option<PreferredJoinChannels>,
+    // List of subbands that haven't already been tried.
+    available_subbands: AvailableSubbands,
+}
+
+impl JoinChannels {
+    /// Select a channel for a join attempt.
+    ///
+    /// ## RNG note:
+    ///
+    /// Uses 1 random number from the RNG.
+    fn select_channel(&mut self, rng: &mut impl RngCore) -> u8 {
+        // Early-return preferred channel if possible
+        if let Some(p) = self.preferred_channels.as_mut().and_then(|p| p.try_get_channel(rng)) {
+            return p;
+        }
+
+        if self.available_subbands.is_empty() {
+            self.available_subbands.reset();
+        }
+
+        // Unwrapping is ok because it should never be empty by this point
+        let subband = self.available_subbands.pop_next().unwrap();
+        8 * subband + (rng.next_u32() % 8) as u8
```

Comment:
this means we will only be joining on the lower 64 channels if I'm reading this right?